airflow/providers/amazon/aws/hooks/rds.py

```diff
@@ -0,0 +1,165 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""Interact with AWS RDS."""
+
+from typing import Any, Callable, Dict, List, Optional, Union
+
+from airflow.exceptions import AirflowException
+from airflow.providers.amazon.aws.hooks.base_aws import AwsBaseHook
+
+
+class RDSHook(AwsBaseHook):
+    """
+    Interact with AWS RDS using the boto3 library.
+    Client methods could be used from hook object, like: hook.create_db_instance()
+
+    Additional arguments (such as ``aws_conn_id``) may be specified and
+    are passed down to the underlying AwsBaseHook.
+
+    All available methods are listed in documentation
+    .. seealso::
+        https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/rds.html
+
+    .. seealso::
+        :class:`~airflow.providers.amazon.aws.hooks.base_aws.AwsBaseHook`
+
+    :param aws_conn_id: The Airflow connection used for AWS credentials.
+    :type aws_conn_id: str
+    """
+
+    def __init__(self, *args, **kwargs) -> None:
+        kwargs["client_type"] = "rds"
+        super().__init__(*args, **kwargs)
+
+    def __getattr__(self, item: str) -> Union[Callable, AirflowException]:
+        client = self.get_conn()
+        if hasattr(client, item):
+            self.log.info("Calling boto3 RDS client's method:", item)
+            return getattr(client, item)
+        else:
+            raise AirflowException("Boto3 RDS client doesn't have a method:", item)
+
+
+class RDSHookAlt(AwsBaseHook):
+    """
+    Interact with AWS RDS, using the boto3 library.
+
+    Additional arguments (such as ``aws_conn_id``) may be specified and
+    are passed down to the underlying AwsBaseHook.
+
+    All unimplemented methods available from boto3 RDS client -> hook.get_conn()
+    .. seealso::
+        https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/rds.html
+
+    .. seealso::
+        :class:`~airflow.providers.amazon.aws.hooks.base_aws.AwsBaseHook`
+
+    :param aws_conn_id: The Airflow connection used for AWS credentials.
+    :type aws_conn_id: str
+    """
+
+    def __init__(self, *args, **kwargs) -> None:
+        kwargs["client_type"] = "rds"
+        super().__init__(*args, **kwargs)
+
+    def describe_db_snapshots(
```

Comment:
hmm yeah i didn't scrutinize the `__getattr__` in the first pass but i see it now

ok so you are just trying to make it so all boto3 methods work on the hook also.

personally i do not think this good.   i see _no_ benefit to doing this because users can just use the `conn` cached property on AwsBaseHook.  (though `client` would have been a better name for the  attr).  in this case we can very easily, without any cost, adhere to the "only one way to do something" principle.

for auto-complete we can add a protocol class for autocomplete on the conn object.  or perhaps we can see if there is a boto3 typing stub library that we could use here (and possibly for other hooks as well)

so to recap my suggestion is, don't use getattr, just direct users to use `conn` or perhaps better yet add property `client` that  returns `self.conn` then add a protocol classs (search the airflow repo for examples) where we stub out the methods for the client

then that's it!

thanks

Reply:
A lot of thanks that clarifying the way to implementation @dstandish 

The solutions below don't work and it's expected 
```
from typing import Type
from botocore.client import BaseClient
self.client: Type[BaseClient]

import boto3
rds_client = boto3.client('rds', 'us-east-1')
self.client: type(rds_client)
```
Did you mean the solution that I started to implement?

Reply:
I've partially implemented the protocol class and, unfortunately, doesn't see sense to continue.
The resulting class will be more than one thousand rows that need to be supported and changed, based on new versions of the boto3 client.
I could finish the implementation, but I need some approval that this code is really needed and won't be deleted during the next reviews.

Reply:
Hi @kazanzhy 

> I could finish the implementation, but I need some approval that this code is really needed and won't be deleted during the next reviews.

I definitely appreciate your concerns @kazanzhy.

I only suggested using `Protocol` as an alternative to duplicating all the client methods on the hook (if the goal is autocomplete).  I wouldn't say it's really necessary and I totally understand your reluctance to add all of this code just for autocomplete.

I just took a look and it seems there is a good library we may be able to leverage here.

Take a look here: https://github.com/vemel/boto3_stubs_docs

I installed with `pip install 'boto3-stubs[s3,ec2,rds]'` and tried out hinting like this:
```python
    def get_conn(self) -> RDSClient:
        return super().get_conn()
```

and it seemed to work.  Maybe this is a good solution here?

What do you think?

Reply:
That's awesome. A lot of thanks :)

I think after this PR we could refactor Amazon provider in the following way:
- add `client` attribute to the hook, which will be typed by some class from boto3_stub
- refactor Operators to replace `hook.get_conn().[method]` by the `hook.client.[method]`

Reply:
Probably you're right.
My motivation to use `client` is that it's more clear and convenient to use `hook.client.some_method()` than `hook.conn.some_method()`.
But I see that redshift operators use `hook.get_conn().some_method()` which is much more unclear in my opinion.

Reply:
I checked your suggestion and autocompletion works only for `hook.conn`.
I think we should not override `get_conn` and I'll use just `hook.conn.some_method()` in my RDS operators

Reply:
> My motivation to use client is that it's more clear and convenient to use hook.client.some_method() than hook.conn.some_method().
But I see that redshift operators use hook.get_conn().some_method() which is much more unclear in my opinion.

Yeah I completely agree.

> I think we should not override get_conn and I'll use just hook.conn.some_method() in my RDS operators

Yeah I'm good with that.